src/lib.rs

```diff
@@ -1256,4 +1342,95 @@ mod tests {
 
         assert_eq!(actual, expected);
     }
+
+    #[test]
+    fn test_outer_join_with_empty_self() {
+        do_test_outer_join(&[], &[1, 2, 2, 3], &[(&1, 0, 1), (&2, 0, 2), (&3, 0, 1)]);
+    }
+
+    #[test]
+    fn test_outer_join_with_empty_other() {
+        do_test_outer_join(&[1, 2, 2, 3], &[], &[(&1, 1, 0), (&2, 2, 0), (&3, 1, 0)]);
+    }
+
+    #[test]
+    fn test_outer_join_with_overlap() {
+        do_test_outer_join(
+            &[1, 2, 2, 3, 3],
+            &[3, 4, 5, 5],
+            &[(&1, 1, 0), (&2, 2, 0), (&3, 2, 1), (&4, 0, 1), (&5, 0, 2)],
+        );
+    }
+
+    fn do_test_outer_join(
+        this: &[usize],
+        other: &[usize],
+        expected_entries: &[(&usize, usize, usize)],
+    ) {
+        let this_hashbag: HashBag<_> = this.iter().cloned().collect();
+        let other_hashbag: HashBag<_> = other.iter().cloned().collect();
+        let expected: HashSet<_> = HashSet::from_iter(expected_entries.iter().cloned());
+        let actual: HashSet<_> = this_hashbag.outer_join(&other_hashbag).collect();
+        assert_eq!(expected, actual);
+    }
```

Comment:
Would appreciate some ideas on how to avoid some code duplication between all these `do_test_<fn>` methods. They're pretty similar, but slightly different, so far I haven't been able to come up with a good way of simplifying this.

I tried to write a generic method that takes `IntoIter`s for the inputs and expected values and a `FnOnce` that transform both inputs into the `actual` value we want to do `assert_eq!` on but that got ugly pretty quickly :/

Reply:
I think it's totally fine to have separate helpers here. Over-generalizing can make the testing code itself hard to follow :)